pythonforandroid/pythonpackage.py

```diff
@@ -0,0 +1,664 @@
+                    return meta_entry.partition(":")[2].strip()
+            if name is None:
+                raise ValueError("failed to obtain package name")
+
+            if extract_type == "namewithpins" and \
+                    dependency.find("==") > 0 and \
+                    dependency.partition("==")[0].lower() == name.lower():
+                return name + "==" + dependency.partition("==")[2].lstrip()
+            return name
+        elif extract_type == "dependencies":
+            requirements = []
+            if os.path.exists(os.path.join(output_folder, 'pyproject.toml')):
+                with open(os.path.join(output_folder, 'pyproject.toml')) as f:
+                    build_sys = pytoml.load(f)['build-system']
+                    if "requires" in build_sys:
+                        requirements += build_sys["requires"]
+
+            # Add requirements from metadata:
+            requirements += [
+                entry.rpartition("Requires-Dist:")[2].strip()
+                for entry in metadata_entries
+                if entry.startswith("Requires-Dist")
+            ]
+
+            return list(set(requirements))  # remove duplicates
+    finally:
+        shutil.rmtree(output_folder)
+
+
+package_name_cache = dict()
+
+
+def get_package_name(dependency,
+                     keep_version_pins=False,
+                     use_cache=True):
+    def timestamp():
+        try:
+            return time.monotonic()
+        except AttributeError:
+            return time.time()  # Python 2.
+    try:
+        value = package_name_cache[dependency]
+        if value[0] + 600.0 > timestamp() and use_cache:
+            return value[1]
+    except KeyError:
+        pass
+    result = _extract_info_from_package(dependency, extract_type="name")
+    package_name_cache[dependency] = (timestamp(), result)
+    return result
+
+
+def get_package_dependencies(package, recursive=False, verbose=False):
+    """ Obtain the dependencies from a package. Please note this
+        function is possibly SLOW, especially if you enable
+        the recursive mode.
+    """
+    packages_processed = set()
+    package_queue = [package]
+    reqs = []
+    reqs_as_names = []
+    while len(package_queue) > 0:
+        current_queue = package_queue
+        package_queue = []
+        for package_dep in current_queue:
+            new_reqs = []
+            if verbose:
+                print("get_package_dependencies: resolving dependecy " +
+                      "to package name: " + str(package_dep))
+            package = get_package_name(package_dep)
+            if package.lower() in packages_processed:
+                continue
+            if verbose:
+                print("get_package_dependencies: processing package: " +
+                      str(package),
+                      )
+                print("get_package_dependencies: Packages seen so far: " +
+                      str(packages_processed),
```

Comment:
minor: don't we find "{}".format() more elegant than str concatenation?

Reply:
yup we do I just keep forgetting :joy:

Reply:
changed :heavy_check_mark: